go/pkg/services/logcollector/server.go

```diff
@@ -0,0 +1,605 @@
+			Error:   err.Error(),
+		}, err
+	}
+
+	// create a child context before calling goroutines, so it won't be canceled
+	logStreamCtx, cancelCtxFunc := mlruncontext.NewDetachedWithCancel(ctx)
+	startedStreamingGoroutine := make(chan bool, 1)
+
+	// stream logs to file
+	go s.startLogStreaming(logStreamCtx, request.RunUID, pod.Name, startedStreamingGoroutine, cancelCtxFunc)
+
+	// add Item to in-memory state
+	if err := s.inMemoryState.AddLogItem(ctx, request.RunUID, request.Selector); err != nil {
+		err := errors.Wrapf(err, "Failed to add run id %s to in memory state", request.RunUID)
+		return &protologcollector.StartLogResponse{
+			Success: false,
+			Error:   err.Error(),
+		}, err
+	}
+
+	<-startedStreamingGoroutine
+
+	return &protologcollector.StartLogResponse{
+		Success: true,
+	}, nil
+}
+
+// GetLogs returns the log file contents of length size from an offset, for a given run id
+func (s *LogCollectorServer) GetLogs(ctx context.Context, request *protologcollector.GetLogsRequest) (*protologcollector.GetLogsResponse, error) {
+	s.Logger.DebugWithCtx(ctx, "Received Get Log request", "request", request)
+
+	// get log file path
+	filePath, err := s.getLogFilePath(request.RunUID)
+	if err != nil {
+		err := errors.Wrapf(err, "Failed to get log file path for run id %s", request.RunUID)
+		return &protologcollector.GetLogsResponse{
+			Success: false,
+			Error:   err.Error(),
+		}, err
+	}
+
+	// read log from file
+	buffer, err := s.readLogsFromFile(ctx, request.RunUID, filePath, request.Offset, request.Size)
+	if err != nil {
+		err := errors.Wrapf(err, "Failed to read logs for run id %s", request.RunUID)
+		return &protologcollector.GetLogsResponse{
+			Success: false,
+			Error:   err.Error(),
+		}, err
+	}
+
+	return &protologcollector.GetLogsResponse{
+		Success: true,
+		Logs:    buffer,
+	}, nil
+}
+
+// startLogStreaming streams logs from a pod and writes them into a file
+func (s *LogCollectorServer) startLogStreaming(ctx context.Context,
+	runUID,
+	podName string,
+	startedStreamingGoroutine chan bool,
+	cancelCtxFunc context.CancelFunc) {
+
+	// in case of a panic, remove this goroutine from the in-memory state, so the
+	// monitoring loop will start logging again for this runUID.
+	defer func() {
+		if err := recover(); err != nil {
+			callStack := debug.Stack()
+			s.Logger.ErrorWithCtx(ctx, "Panic caught while creating function",
+				"err", err,
+				"stack", string(callStack))
+
+			// remove this goroutine from in-memory state
+			if err := s.inMemoryState.RemoveLogItem(runUID); err != nil {
+				s.Logger.WarnWithCtx(ctx, "Failed to remove item from in memory state")
+			}
+
+			// cancel all other goroutines spawned from this one
+			cancelCtxFunc()
+		}
+	}()
+
+	// signal "main" function that goroutine is up
+	startedStreamingGoroutine <- true
+
+	// create a log file to the pod
+	logFilePath := s.resolvePodLogFilePath(runUID, podName)
+	if err := common.EnsureFileExists(logFilePath); err != nil {
```

Comment:
🐛 
if you have an issue here to create the log file. your inmemory/persistency log collector says "in progress" while.. you are not doing anything